src/compressor/QatAccel.cc

```diff
@@ -12,58 +12,146 @@
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_compressor
+#undef dout_prefix
+#define dout_prefix _prefix(_dout)
+
+static std::ostream& _prefix(std::ostream* _dout)
+{
+  return *_dout << "QatAccel: ";
+}
+// -----------------------------------------------------------------------------
+
 void QzSessionDeleter::operator() (struct QzSession_S *session) {
-  if (NULL != session->internal) {
-    qzTeardownSession(session);
-    qzClose(session);
-  }
+  qzTeardownSession(session);
   delete session;
 }
 
 /* Estimate data expansion after decompression */
 static const unsigned int expansion_ratio[] = {5, 20, 50, 100, 200};
 
-QatAccel::QatAccel() {
-  session.reset(new struct QzSession_S);
-  memset(session.get(), 0, sizeof(struct QzSession_S));
-}
-
-QatAccel::~QatAccel() {}
-
-bool QatAccel::init(const std::string &alg) {
-  QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+static bool get_qz_params(const std::string &alg, QzSessionParams_T &params) {
   int rc;
-
   rc = qzGetDefaults(&params);
   if (rc != QZ_OK)
     return false;
   params.direction = QZ_DIR_BOTH;
-  if (alg == "zlib")
+  params.is_busy_polling = true;
+  if (alg == "zlib") {
     params.comp_algorithm = QZ_DEFLATE;
-  else
+    params.data_fmt = QZ_DEFLATE_GZIP_EXT;
+  }
+  else {
+    // later, there also has lz4.
     return false;
+  }
 
   rc = qzSetDefaults(&params);
   if (rc != QZ_OK)
       return false;
+  return true;
+}
 
+static bool setup_session(QatAccel::session_ptr &session, QzSessionParams_T &params) {
+  int rc;
   rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+  if (rc != QZ_OK && rc != QZ_DUPLICATE)
     return false;
-
   rc = qzSetupSession(session.get(), &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(session.get());
-    qzClose(session.get());
+  if (rc != QZ_OK) {
     return false;
   }
+  return true;
+}
+
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
+  cached_session_t(QatAccel* accel, QatAccel::session_ptr&& sess)
+    : accel{accel}, session{std::move(sess)} {}
+
+  ~cached_session_t() {
+    std::scoped_lock lock{accel->mutex};
+    // if the cache size is still under its upper bound, the current session is put into
+    // accel->sessions. otherwise it's released right
+    uint64_t sessions_num = g_ceph_context->_conf.get_val<uint64_t>("qat_compressor_session_max_number");
+    if (accel->sessions.size() < sessions_num) {
+      accel->sessions.push_back(std::move(session));
+    }
+  }
 
+  struct QzSession_S* get() {
+    assert(static_cast<bool>(session));
+    return session.get();
+  }
+
+  QatAccel* accel;
+  QatAccel::session_ptr session;
+};
+
+QatAccel::session_ptr QatAccel::get_session() {
+  std::scoped_lock lock{mutex};
+  if (!sessions.empty()) {
+    auto session = std::move(sessions.back());
+    sessions.pop_back();
+    return session;
+  }
```

Comment:
is it safe to drop the lock at this point?

Reply:
The lock can not be removed, but it can be limited only on sessions not allocating a new session.